// @qiwi/substrate-std/target/es6/stack.js
"use strict";
/** @module @qiwi/substrate-std */
/** */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stack = void 0;
/**
 * Standard stack implementation
 */
class Stack {
    constructor(...args) {
        // super(...args)
        this._storage = [...args];
    }
    get(index) {
        return this._storage[+index];
    }
    push(...items) {
        this._storage.push(...items);
        return items[items.length - 1];
    }
    pop() {
        return this._storage.pop();
    }
    shift() {
        return this._storage.shift();
    }
    unshift(...items) {
        this._storage.unshift(...items);
        return items[0];
    }
    get size() {
        return this._storage.length;
    }
    remove(index) {
        const idx = +index;
        const found = this._storage[+idx];
        this._storage.splice(idx, 1);
        return found;
    }
    last() {
        return this._storage[this.size - 1];
    }
    first() {
        return this._storage[0];
    }
    toArray() {
        return [...this._storage];
    }
    clear() {
        this._storage.length = 0;
    }
    indexOf(item) {
        return this._storage.indexOf(item);
    }
    add(index, item) {
        this._storage[+index] = item;
        return item;
    }
    isEmpty() {
        return this.size > 0;
    }
}
exports.Stack = Stack;
